clippy_lints/src/ignored_unit.rs

```diff
@@ -0,0 +1,54 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use hir::PatKind;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `_` in patterns of type `()`.
+    ///
+    /// ### Why is this bad?
+    /// Matching with `()` explicitely instead of `_` outlines
+    /// the fact that the pattern contains no data.
+    ///
+    /// ### Example
+    /// ```rust
+    /// match std::fs::create_dir("tmp-work-dir") {
+    ///    Ok(_) => println!("Working directory created"),
+    ///    Err(s) => eprintln!("Could not create directory: {s}"),
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// match std::fs::create_dir("tmp-work-dir") {
+    ///    Ok(()) => println!("Working directory created"),
+    ///    Err(s) => eprintln!("Could not create directory: {s}"),
+    /// }
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub IGNORED_UNIT,
+    pedantic,
+    "suggest replacing `_` by `()` in patterns where appropriate"
+}
+declare_lint_pass!(IgnoredUnit => [IGNORED_UNIT]);
+
+impl<'tcx> LateLintPass<'tcx> for IgnoredUnit {
+    fn check_pat(&mut self, cx: &LateContext<'tcx>, pat: &'tcx hir::Pat<'tcx>) {
+        pat.walk(|pat| {
+            if matches!(pat.kind, PatKind::Wild) && cx.tcx.typeck(pat.hir_id.owner.def_id).pat_ty(pat).is_unit() {
+                span_lint_and_sugg(
+                    cx,
+                    IGNORED_UNIT,
+                    pat.span,
+                    "matching over `()` is more explicit",
+                    "use `()` instead of `_`",
+                    String::from("()"),
+                    Applicability::MachineApplicable,
+                );
+            }
+            true
```

Comment:
nit: If removing `pat.walk` doesn't work, this should use `walk_always` instead.